### PR TITLE
Hash collisions redmine7912

### PR DESCRIPTION
**Pull request by jimis**

### CODE REVIEW

**Comment by jimis** [comment]
trigger build

**Comment by jimis** [comment]
@kacf please review

**Comment by jimis** [comment]
First build is green, triggering again because I rebased with a couple of cosmetic changes.

**Comment by jimis** [comment]
trigger build

**Comment by jimis** [comment]
All green, I'll now amend a commit with an acceptance test.

**Comment by jimis** [comment]
Acceptance test added, tested failing before the patch and passing afterwards.

**Comment by jimis** [comment]
trigger build

**Comment by jimis** [comment]
all green

**Comment by jimis** [comment]
@kacf please merge if it looks ok after you have a look

**Comment by jimis** [comment]
Fixed small memory leak, squashing it to the original commit.

**Comment by jimis** [comment]
trigger build

File: libpromises/class.c [hunk 1]
```diff
@@ -244,11 +242,20 @@ ClassTableIterator *ClassTableIteratorNew(const ClassTable *table, const char *n
...
     {
+        Class *cls = keyvalue->value;
+
+        /* Make sure we never store "default" as namespace in the ClassTable,
+         * instead we have always ns==NULL in that case. */
+        CF_ASSERT_FIX(cls->ns == NULL ||
+                      strcmp(cls->ns, "default") != 0,
+                      (cls->ns = NULL),
+                      "Class table contained \"default\" namespace,"
+                      " should never happen!");
```
Comment (kacf) [inline libpromises/class.c]: I'm confused about this part. Here you say that you don't store the namespace, but you *do* store the namespace specifically in `ClassTablePut()`, by using `fullname`. Am I missing something?

**Comment by kacf** [comment]
Exactly the type of fix I was hoping for! Just two issues:

* The namespace storing issue I mentioned in the code comment.
* Changelog entry is missing.

File: libpromises/class.c (hunk 1 above)
Comment (jimis) [inline libpromises/class.c]: On Wed, Mar 2, 2016 at 9:44 AM, Kristian Amlie <notifications@github.com>
wrote:

> In libpromises/class.c
> <https://github.com/cfengine/core/pull/2506#discussion_r54693749>:
>
> >      {
> > +        Class *cls = keyvalue->value;
> > +
> > +        /* Make sure we never store "default" as namespace in the ClassTable,
> > +         * instead we have always ns==NULL in that case. */
> > +        CF_ASSERT_FIX(cls->ns == NULL ||
> > +                      strcmp(cls->ns, "default") != 0,
> > +                      (cls->ns = NULL),
> > +                      "Class table contained \"default\" namespace,"
> > +                      " should never happen!");
>
> I'm confused about this part. Here you say that you don't store the
> namespace, but you *do* store the namespace specifically in
> ClassTablePut(), by using fullname. Am I missing something?
>
>
"fullname" is the key, in this key-value store, and is indeed always
including the namespace. However this assertion ensures that the *value* of
the key-value store, i.e. struct Class, is in the correct form.

BTW this assertion is rather expensive. I'm contemplating on introducing
something like CF_DEBUG_ASSERT that completely disappears ifdef NDEBUG.

**Comment by jimis** [comment]
Added changelog message and rebased.

**Comment by jimis** [comment]
Thanks for reviewing @kacf. I'm merging to continue with other fixes on top of it.

**Pull request merged**